src/tri.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2014-2024 bluss and ndarray developers.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use core::cmp::{max, min};
+
+use num_traits::Zero;
+
+use crate::{Array, ArrayBase, Axis, Data, Dimension, IntoDimension, RemoveAxis, Zip};
+
+impl<S, A, D> ArrayBase<S, D>
+where
+    S: Data<Elem = A>,
+    D: Dimension + RemoveAxis,
+    A: Clone + Zero,
+    D::Smaller: Copy,
+{
+    /// Upper triangular of an array.
+    ///
+    /// Return a copy of the array with elements below the *k*-th diagonal zeroed.
+    /// For arrays with `ndim` exceeding 2, `triu` will apply to the final two axes.
+    ///
+    /// See also [`ArrayBase::tril`]
+    ///
+    /// ```
+    /// use ndarray::array;
+    ///
+    /// let arr = array![[1, 2, 3], [4, 5, 6], [7, 8, 9]];
+    /// let res = arr.triu(0);
+    /// assert_eq!(res, array![[1, 2, 3], [0, 5, 6], [0, 0, 9]]);
+    /// ```
+    pub fn triu(&self, k: isize) -> Array<A, D> {
+        let mut res = Array::zeros(self.raw_dim());
+        Zip::indexed(self.rows())
+            .and(res.rows_mut())
+            .for_each(|i, src, mut dst| {
+                let row_num = i.into_dimension().last_elem();
+                let lower: isize = max(row_num as isize + k, 0);
+                dst.slice_mut(s![lower..]).assign(&src.slice(s![lower..]));
+            });
+
+        res
+    }
+
+    /// Lower triangular of an array.
+    ///
+    /// Return a copy of the array with elements above the *k*-th diagonal zeroed.
+    /// For arrays with `ndim` exceeding 2, `tril` will apply to the final two axes.
+    ///
+    /// See also [`ArrayBase::triu`]
+    ///
+    /// ```
+    /// use ndarray::array;
+    ///
+    /// let arr = array![[1, 2, 3], [4, 5, 6], [7, 8, 9]];
+    /// let res = arr.tril(0);
+    /// assert_eq!(res, array![[1, 0, 0], [4, 5, 0], [7, 8, 9]]);
+    /// ```
+    pub fn tril(&self, k: isize) -> Array<A, D> {
+        let mut res = Array::zeros(self.raw_dim());
```

Comment:
This will create a c-order array even if the input is a f-order array. I think calling `clone` is better here.

Reply:
Ah actually this is something I was unclear about - is there a way to create an array of zeros with the same shape *and* order? Or are you saying I should `.clone` the input and fill it with zeros, instead?

Reply:
I think `Array::zeros(self.raw_dim().set_f(is_layout_f(&self.dim, &self.strides)))` is doing what I want, just don't know if there's another way.

Reply:
No, you're right. `clone()` is sexier but it would force you to fill the array with 0 after cloning, which is wasteful.

I don't know if the maintainers usually try making optimized version for f-order arrays, but one could make a fast version for f-order arrays by calling `columns` instead of `rows`. Feels like code duplication though. I'm not asking you to do it, I'm just wondering.